pkg/apis/seedmanagement/types_managedseed.go

```diff
@@ -0,0 +1,182 @@
+	// Env is the list of environment variables to set in the gardenlet container.
+	Env []corev1.EnvVar
+	// VPA specifies whether to enable VPA for gardenlet. Defaults to false.
+	VPA *bool
+	// ImageVectorOverwrite is the gardenlet image vector overwrite.
+	// More info: https://github.com/gardener/gardener/blob/master/docs/deployment/image_vector.md.
+	ImageVectorOverwrite string
+	// ComponentImageVectorOverwrites is a list of image vector overwrites for components deployed by gardenlet.
+	// More info: https://github.com/gardener/gardener/blob/master/docs/deployment/image_vector.md.
+	ComponentImageVectorOverwrites string
+}
+
+// Image specifies container image parameters.
+type Image struct {
+	// Repository is the image repository.
+	Repository string
+	// Tag is the image tag.
+	Tag string
+	// PullPolicy is the image pull policy. One of Always, Never, IfNotPresent.
+	PullPolicy corev1.PullPolicy
+}
+
+// GardenConnectionBootstrap describes a mechanism for bootstrapping gardenlet connection to the Garden cluster.
+type GardenConnectionBootstrap string
+
+const (
+	// GardenConnectionBootstrapServiceAccount means that a service account should be used for bootstrapping gardenlet connection to the Garden cluster.
+	GardenConnectionBootstrapServiceAccount GardenConnectionBootstrap = "ServiceAccount"
+	// GardenConnectionBootstrapToken means that a bootstrap token should be used for bootstrapping gardenlet connection to the Garden cluster.
+	GardenConnectionBootstrapToken GardenConnectionBootstrap = "BootstrapToken"
+)
+
+// SeedConnection describes a mechanism for gardenlet connection to the Seed cluster.
+type SeedConnection string
+
+const (
+	// SeedConnectionServiceAccount means that a service account should be used for gardenlet connection to the Seed cluster.
+	SeedConnectionServiceAccount SeedConnection = "ServiceAccount"
+)
+
+// APIServer specifies certain kube-apiserver parameters of the Shoot that will be registered as Seed.
+type APIServer struct {
+	// Replicas is the number of kube-apiserver replicas. Defaults to 3.
+	Replicas *int32
+	// Autoscaler specifies certain kube-apiserver autoscaler parameters, such as the minimum and maximum number of replicas.
+	Autoscaler *APIServerAutoscaler
+}
+
+// APIServerAutoscaler specifies certain kube-apiserver autoscaler parameters of the Shoot that will be registered as Seed.
+type APIServerAutoscaler struct {
+	// MinReplicas is the minimum number of kube-apiserver replicas. Defaults to min(3, MaxReplicas).
+	MinReplicas *int32
+	// MaxReplicas is the maximum number of kube-apiserver replicas. Defaults to 3.
+	MaxReplicas *int32
+}
+
+// ManagedSeedStatus is the status of a ManagedSeed.
+type ManagedSeedStatus struct {
+	// LastOperation holds information about the last operation on the ManagedSeed.
+	LastOperation *gardencorev1beta1.LastOperation
+	// LastError holds information about the last occurred error during an operation.
+	LastError *gardencorev1beta1.LastError
```

Comment:
I think, the reconciliation of a `ManagedSeed` is quite simple and doesn't involve many interconnected actions/operations, so we don't need a `LastOperation` and `LastError` field in this case, IMO.
`LastOperation` is an indicator of long running reconciliation loops, which is basically contradicting to kubernetes conventions of having simple and fast reconciliation loops, e.g. see the Deployment controller: it can also take quite a long time for a given rollout to complete, nevertheless a single reconciliation run of the deployment controller takes only a few milliseconds.

I would suggest to model the status rather in conditions (e.g. one could be `ShootReconciled` indicating, that the shoot has been reconciled at least once, and another could be `GardenletDeployed`/`GardenletReady` indicating that the `ManagedSeed` controller has successfully deployed the gardenlet). WDYT?

Reply:
Sounds good, I will think about it and either adapt the implementation or reply here if I have concerns. Note however that reconciliation of ManagedSeeds is not necessarily always that quick - especially during deletion, it involves deleting Gardenlet, deleting all resources referencing the Seed (backup entries, etc.), only after that deleting the Seed secrets (since otherwise they may be gone too early and the entire process is stuck, I experienced this a lot during testing), and finally deleting the `garden` namespace. I am not sure that `LastOperation` or `LastError` would not actually prove useful in such cases.

Reply:
> Note however that reconciliation of ManagedSeeds is not necessarily always that quick - especially during deletion, it involves deleting Gardenlet, deleting all resources referencing the Seed (backup entries, etc.)

Yes, the process of reconciling/deleting a ManagedSeed can indeed take quite a while. But it's the same for rolling a huge Deployment (e.g. with long wait times in between). However the Deployment controller always only checks, what it should do and if the observed state is ready for the next step, but it never waits (i.e. blocks a worker goroutine).
I think, the ManagedSeed controller is actually quite similar. It can check if the Shoot is already ready, if not requeue, and otherwise proceed with the gardenlet deployment. In any case, it can update status fields / conditions accordingly, which would be sufficient to transport the current state of the object.
Do you get, what I am saying?

Reply:
Yes, I think I do. I guess I should remove the `LastOperation` / `LastError` in this case.